device/core/src/module_client.ts

```diff
@@ -126,11 +131,76 @@ export class ModuleClient extends InternalClient {
     });
   }
 
+  /**
+   * Closes the client and the associated transport connection and resources.
+   *
+   * @param closeCallback callback that shall be called with either an error or a `results.Disconnected` object when the transport has successfully closed its connection.
+   */
   close(closeCallback?: (err?: Error, result?: results.Disconnected) => void): void {
     this._transport.removeListener('disconnect', this._moduleDisconnectHandler);
     super.close(closeCallback);
   }
 
+  /**
+   * Invokes a method on a downstream device or on another module on the same Edge device. Please note that this feature only works when
+   * the module is being run as part of an Edge device.
+   *
+   * @param deviceId      target device identifier
+   * @param moduleId      target module identifier on the device identified with the `deviceId` argument
+   * @param methodParams  parameters of the direct method call
+   * @param callback      callback that will be invoked either with an Error object or the result of the method call.
+   */
+  invokeMethod(deviceId: string, methodParams: MethodParams, callback: MethodCallback): void;
+  invokeMethod(deviceId: string, moduleId: string, methodParams: MethodParams, callback: MethodCallback): void;
+  invokeMethod(deviceId: string, moduleIdOrMethodParams: string | MethodParams, methodParamsOrCallback: MethodParams | MethodCallback, callback?: MethodCallback): void {
+    /*Codes_SRS_NODE_MODULE_CLIENT_16_093: [`invokeMethod` shall throw a `ReferenceError` if the `deviceId` argument is falsy.]*/
+    if (!deviceId) {
+      throw new ReferenceError('deviceId cannot be \'' + deviceId + '\'');
+    }
+    /*Codes_SRS_NODE_MODULE_CLIENT_16_094: [`invokeMethod` shall throw a `ReferenceError` if the `moduleIdOrMethodParams` argument is falsy.]*/
+    if (!moduleIdOrMethodParams) {
+      if (typeof methodParamsOrCallback === 'object') {
+        throw new ReferenceError('moduleId cannot be \'' + moduleIdOrMethodParams + '\'');
+      } else {
+        throw new ReferenceError('methodParams cannot be \'' + moduleIdOrMethodParams + '\'');
+      }
```

Comment:
for maintainability and readability, I would collapse these into one exception "second parameter cannot be 'blah'.  Otherwise, it could be confusing to the caller because you're trying to infer the meaning of the second parameter based on the type of the third parameter.  
This particular group is functions is just hard to read and grok.  Maybe not too important since it's just parameter checking, but I would need to take out a piece of paper just to make sure it works correctly and that's a bad sign.

Reply:
Maybe this would be easier if you moved the overloads (with their confusing names) into actualX values and then verified those instead.  I think that would be way way easier to understand.  

---
In reply to: [197229768](https://github.com/Azure/azure-iot-sdk-node/pull/322#discussion_r197229768) [](ancestors = 197229768)